### ISSUE TITLE
1 gpu device is ok, bug 2 gpu devices error and 8 gpu devices timeout?

### ISSUE THREAD

**Issue by abmybgx** [kata-containers/kata-containers#7567]
# Description of problem

It's ok for 1 device.
ctr --debug -n k8s.io run --runtime "io.containerd.kata.v2" --device /dev/vfio/47 --rm -t registry.xxx.com/xxx/ubuntu18.04-py3.7-cuda11.2-cudnn8-torch1.8-openmmlab:b06131830 test2
<img width="589" alt="image" src="https://github.com/kata-containers/kata-containers/assets/34848440/b745cdf8-7c48-4473-8d15-78bc5bc63a25">

Error for 2 devices.
ctr --debug -n k8s.io run --runtime "io.containerd.kata.v2" --device /dev/vfio/47 --device /dev/vfio/125 --rm -t registry.xxx.com/xxx/ubuntu18.04-py3.7-cuda11.2-cudnn8-torch1.8-openmmlab:b06131830 test2
```
nvidia-container-cli: container error: open failed: /proc/1125/root/run/kata-containers/test2/rootfs: no such file or directory
```

Timeout for 8 devices.
ctr --debug -n k8s.io run --runtime "io.containerd.kata.v2" --device /dev/vfio/125 --device /dev/vfio/129 --device /dev/vfio/168 --device /dev/vfio/172 --device /dev/vfio/54 --device /dev/vfio/76 --device /dev/vfio/83  --device /dev/vfio/47 --rm -t registry.xxx.com/xxx/ubuntu18.04-py3.7-cuda11.2-cudnn8-torch1.8-openmmlab:b06131830 test2
```
nvidia-container-cli: initialization error: driver rpc error: timed out
```

# Expected result

2 gpu or more could be ok.

# Actual result

only 1 gpu could be ok

# Further information

[kata.log](https://github.com/kata-containers/kata-containers/files/12279032/kata.log)

**Comment by abmybgx** [comment]
Seems like https://github.com/NVIDIA/libnvidia-container/issues/19 ?

**Comment by zvonkok** [comment]
Which version of Kata are you using? 
Can you also share your configuration.toml?

**Comment by abmybgx** [comment]
> Contributor

Thanks for ur reply, pls check blow.

![image](https://github.com/kata-containers/kata-containers/assets/34848440/e66dda14-0ee1-4286-8d44-5f02c680b4d9)

```
[hypervisor.qemu]
path = "/opt/kata/bin/qemu-system-x86_64"
kernel = "/opt/kata/share/kata-containers/vmlinux-nvidia-gpu.container"
image = "/opt/kata/share/kata-containers/kata-containers.img"
machine_type = "q35"
rootfs_type="ext4"
enable_annotations = ["enable_iommu"]
valid_hypervisor_paths = ["/opt/kata/bin/qemu-system-x86_64"]
kernel_params = ""
firmware = ""
firmware_volume = ""
machine_accelerators=""
cpu_features="pmu=off"
default_vcpus = -1
default_maxvcpus = 0
default_bridges = 1
default_memory = 6000
default_maxmemory = 0
disable_block_device_use = false
shared_fs = "virtio-fs"
virtio_fs_daemon = "/opt/kata/libexec/virtiofsd"
valid_virtio_fs_daemon_paths = ["/opt/kata/libexec/virtiofsd"]
virtio_fs_cache_size = 0
virtio_fs_queue_size = 1024
virtio_fs_extra_args = ["--thread-pool-size=1", "--announce-submounts"]
virtio_fs_cache = "auto"
block_device_driver = "virtio-scsi"
block_device_aio = "io_uring"
enable_iothreads = false
enable_vhost_user_store = false
vhost_user_store_path = "/var/run/kata-containers/vhost-user"
valid_vhost_user_store_paths = ["/var/run/kata-containers/vhost-user"]
vhost_user_reconnect_timeout_sec = 0
valid_file_mem_backends = [""]
pflashes = []
hotplug_vfio_on_root_bus = true
pcie_root_port = 8
valid_entropy_sources = ["/dev/urandom","/dev/random",""]
guest_hook_path = "/usr/share/oci/hooks"
disable_selinux=true
disable_guest_selinux=true
[factory]
[agent.kata]
enable_debug = true
enable_tracing = true
kernel_modules=[]
debug_console_enabled = true
[runtime]
enable_debug = true
internetworking_model="tcfilter"
disable_guest_seccomp=true
sandbox_cgroup_only=false
static_sandbox_resource_mgmt=false
sandbox_bind_mounts=[]
vfio_mode="guest-kernel"
disable_guest_empty_dir=false
experimental=[]
```

**Comment by abmybgx** [comment]
I just tried version 2.5.2, there were different error

```
ctr: failed to create shim task: ETIMEDOUT: Connection timed out: unknown
```

**Comment by zvonkok** [comment]
Can you try the 3.2.0-something this has several updates for handling GPUs.

**Comment by zvonkok** [comment]
What kind of GPU do you have?

**Comment by abmybgx** [comment]
> What kind of GPU do you have?

A100 * 8

**Comment by abmybgx** [comment]
@zvonkok hi, seems like that was because of my prestart hook timeout, could I set the wait time for my prestart hook and how?

```
{"msg":"hook /usr/share/oci/hooks/prestart/nvidia-container-toolkit.sh failed: timeout","level":"ERRO","ts":"2023-08-11T07:21:49.644567251Z","version":"0.1.0","name":"kata-agent","subsystem":"rpc","source":"agent","pid":"979"}
```

**Comment by abmybgx** [comment]
solved by rebuild kata-agent for version 3.1.3，increased below value.

```
const DEFAULT_HOOK_TIMEOUT_SEC: i32 = 10;
```

**Comment by hfyeh** [comment]
For those who want to change the prestart hook's default timeout, this line is what you're looking for.

https://github.com/kata-containers/kata-containers/blob/c018a1cc61d70040a5d6dfc69e8ec9e02f9427c8/src/libs/kata-sys-util/src/hooks.rs#L19

Specifically, `const DEFAULT_HOOK_TIMEOUT_SEC` is the default value.
